File: matlab_code/hb_counter_autocorr.m
function [heartBeats, avg_hr, debug] = hb_counter_autocorr(temporal_mean, fr, firstSample, window_size, overlap_ratio, minPeakDistance, threshold)
	%Only 1 window
	%window_size = length(temporal_mean) - firstSample;
	
	
	% Step 1: Calculate the window-based autocorrelation of the signal stream
	windowStart = firstSample; %Int
	autocorrelation = []; %Double vector
	last_segment_end_value = 0; %Double
	while windowStart < length(temporal_mean)
		% Window to calculate the autocorrelation for
		windowEnd = min(windowStart + window_size - 1, length(temporal_mean));
		segment = temporal_mean(windowStart : windowEnd); %Double vector
	
		% Calculate the autocorrelation for the current window
		local_autocorr = filter(segment - mean(segment), 1, segment - mean(segment)); %Double vector
		local_autocorr = local_autocorr - local_autocorr(1) + last_segment_end_value;
		
		% Define the segment length
		% a. Shine-step-counting style
		[~, max_peak_locs] = findpeaks(local_autocorr, 'MINPEAKDISTANCE', minPeakDistance);
		%[Double vector, Int vector]
		
		if isempty(max_peak_locs)
			segment_length = length(segment); %Int
		else
			[~, min_peak_locs] = findpeaks(-local_autocorr, 'MINPEAKDISTANCE', minPeakDistance); %Int vector
			
			if isempty(min_peak_locs)
				segment_length = round((max(max_peak_locs) + window_size) / 2); %Int
			else
				segment_length = round((max(min_peak_locs) + max(max_peak_locs)) / 2); %Int
			end
		end
		
		% b. Equal-step progression
		%segment_length = length(segment);
		
		% Record the autocorrelation for the current window
		autocorrelation(windowStart : windowStart + length(local_autocorr) - 1) = local_autocorr;
		
		% Define the start of the next window
		windowStart = windowStart + round((1 - overlap_ratio) * segment_length);
		last_segment_end_value = autocorrelation(windowStart - 1);
	end
	
	% Step 2: perform peak-counting on the autocorrelation stream
	windowStart = firstSample;
	heartBeats = []; %Tx2 array: col 1 == double, col 2 == int
	heartRates = []; %Double vector
	while windowStart < length(autocorrelation)
		windowEnd = min(windowStart + window_size - 1, length(autocorrelation));
		segment = autocorrelation(windowStart : windowEnd);
	
		[max_peak_strengths, max_peak_locs] = findpeaks(segment, 'MINPEAKDISTANCE', minPeakDistance);
		
		% Define the segment length
		% a. Shine-step-counting style
		if isempty(max_peak_locs)
			segment_length = length(segment); %Int
			
			heartRates(windowStart : windowStart + segment_length - 1) = zeros(1, segment_length);
		else
			[~, min_peak_locs] = findpeaks(-segment, 'MINPEAKDISTANCE', minPeakDistance, 'THRESHOLD', threshold); %Int vector
			
			if isempty(min_peak_locs)
				segment_length = round((max(max_peak_locs) + window_size) / 2); %Int
			else
				segment_length = round((max(min_peak_locs) + max(max_peak_locs)) / 2); %Int
			end
			
			heartRates(windowStart : windowStart + segment_length - 1) = ones(1, segment_length) * length(max_peak_locs) / sum(isfinite(segment)) * fr;
		end
		
		% b. Equal-step progression
		%segment_length = length(segment);
		
		% Record all beats in the window, even if there are duplicates
		heartBeats = [heartBeats; [max_peak_strengths', (windowStart - 1 + max_peak_locs)']];		
		
		% Define the start of the next window
		windowStart = windowStart + round((1 - overlap_ratio) * segment_length);
	end
	
	% Prune the beats counted to include only unique ones
	heartBeats = unique(heartBeats, 'rows', 'stable');
	
	% Calculate the average HR for the whole stream
	if ~isempty(heartBeats)
		%avg_hr = round(size(heartBeats, 1) / length(heartRates(firstSample : end)) * fr * 60); %Double
		
		number_of_relevant_frames = length(heartRates(firstSample : end)) - sum(~isfinite(temporal_mean(firstSample : end))); %Int
		if number_of_relevant_frames ~= 0
			relevant_time = number_of_relevant_frames / (fr * 60); %Double
			avg_hr = size(heartBeats, 1) / relevant_time; %Double
		else
			avg_hr = 0;
		end
	else
		avg_hr = 0;
	end
	
	debug.heartRates = heartRates;
	debug.autocorrelation = autocorrelation;

File: MisfitHeartRate/MisfitHeartRate/algorithms/HeartRateCal/hb_counter_autocorr.cpp
//
//  hb_counter_autocorr.cpp
//  MisfitHeartRate
//
//  Created by Thanh Le on 7/2/14.
//  Copyright (c) 2014 misfit. All rights reserved.
//

#include "hb_counter_autocorr.h"


namespace MHR {
    vector<int> hb_counter_autocorr(vector<double> &temporal_mean, double fr, int firstSample,
                            int window_size, double overlap_ratio, double minPeakDistance, hrDebug& debug)
    {
        // Step 1: calc the window-based autocorrelation of the signal stream
        
        int windowStart = firstSample-1;
        vector<double> autocorrelation;
        double lastSegmentEndVal = 0;
        bool isFirstSegment = true;

        while(windowStart < (int)temporal_mean.size() - 1){
            
            vector<double> segment;
            vector<double> max_peak_strengths, min_peak_strengths;
            vector<int> max_peak_locs, min_peak_locs;
            
            int segment_length;
            
            //Window to calculate the autocorrelation
            int windowEnd = min(windowStart + window_size, (int)temporal_mean.size());
            for(int i = windowStart; i < windowEnd; ++i)
                segment.push_back(temporal_mean[i]);
            
            //calc mean and get segment = segment - mean
//            double mean = mean(segment);
//            for(int i=0; i<(int) segment.size(); ++i) segment[i]-=mean;
            
            //get the reverse vector of segment
//            vector<double> rev_segment=segment;
//            reverse(rev_segment.begin(), rev_segment.end());
            
            //Calculate the autocorrelation for the current window
            vector<double> local_autocorr = corr_linear(segment, segment);
            double tmp = local_autocorr[0] - lastSegmentEndVal;
            for(int i = 0, sz = (int)local_autocorr.size(); i < sz; ++i)
                local_autocorr[i] -= tmp;
            
            //Define the segment length
            
            // a. Shine-step-counting style
            findpeaks(local_autocorr, minPeakDistance, 0, max_peak_strengths, max_peak_locs);
            
            if(max_peak_locs.empty()){
                segment_length = (int)segment.size();
            }else{
                for(int i=0; i<(int) local_autocorr.size(); ++i) local_autocorr[i] = -local_autocorr[i];
                findpeaks(local_autocorr, minPeakDistance, 0, min_peak_strengths, min_peak_locs);
                
                if(min_peak_locs.empty()){
                    segment_length = round((*max_element(max_peak_locs.begin(), max_peak_locs.end()) + window_size)/2.0 + 1); //round
                    segment_length = min(segment_length, (int)segment.size());
                }else{
                    segment_length = round((*max_element(max_peak_locs.begin(), max_peak_locs.end())
                                      + *max_element(min_peak_locs.begin(), min_peak_locs.end()))/2.0 + 1); //round
                }
                for(int i=0; i<(int) local_autocorr.size(); ++i) local_autocorr[i] = -local_autocorr[i];
            }
            
            
            // b. Equal-step progression
            // segment_length = window_size
            
            // c. autocorrelation
            int windowUpdate = int((1-overlap_ratio)*segment_length+0.5+1e-9);
            if (isFirstSegment) {
                for (int i = 0; i < windowStart; ++i)
                    autocorrelation.push_back(0);
                isFirstSegment = false;
            }
            for(int i = 0, sz = min((int)local_autocorr.size(), windowUpdate); i < sz; ++i)
                autocorrelation.push_back(local_autocorr[i]);

            // Define the start of the next window
            windowStart = windowStart + windowUpdate;
            lastSegmentEndVal = autocorrelation[(int)autocorrelation.size() - 1];
        }
               
//        if (DEBUG_MODE) {
//            String path = _outputPath + "6_autocorrelation.txt";
//            FILE *file = fopen(path.c_str(), "w");
//            fprintf(file, "fr = %lf\nfirstSample = %d\nwindow_size = %d\n", fr, firstSample, window_size);
//            fprintf(file, "overlap_ratio = %lf\nminPeakDistance = %lf\n", overlap_ratio, minPeakDistance);
//            fclose(file);
//            writeVector(autocorrelation, _outputPath + "6_autocorrelation.txt", true);
//        }
        
        // Step 2: perform peak-counting on the autocorrelation stream
        windowStart = firstSample-1;
        vector<pair<double, int>> heartBeats;
        vector<double> heartRates;
        isFirstSegment = true;
        while(windowStart < (int)autocorrelation.size() - 1){
            
            vector<double> segment;
            vector<double> max_peak_strengths, min_peak_strengths;
            vector<int> max_peak_locs, min_peak_locs;
            int segment_length;
            
            int windowEnd = min(windowStart + window_size, (int)autocorrelation.size());
            for(int i = windowStart; i < windowEnd; ++i)
                segment.push_back(autocorrelation[i]);
            
            //Count the number of peaks in this window
            findpeaks(segment, minPeakDistance, 0, max_peak_strengths, max_peak_locs);
            
            //Define the segment length
            // a. Shine-step-counting style
            if(max_peak_locs.empty()){
                segment_length = (int)segment.size();
            }else{
                for(int i=0; i<(int) segment.size(); ++i) segment[i]=-segment[i];
                findpeaks(segment, minPeakDistance, 0, min_peak_strengths, min_peak_locs);
                if(min_peak_locs.empty()){
                    segment_length = round((*max_element(max_peak_locs.begin(), max_peak_locs.end()) + window_size)/2.0 + 1); //round
                    segment_length = min(segment_length, (int)segment.size());
                }else{
                    segment_length = round((*max_element(max_peak_locs.begin(), max_peak_locs.end())
                                      + *max_element(min_peak_locs.begin(), min_peak_locs.end()))/2.0 + 1); //round
                }
                for(int i=0; i<(int) segment.size(); ++i) segment[i]=-segment[i];
            }
            
            // b. Equal_step progression
            // segment_length = window_size;
            
            // Record all beats in the window, even if there are duplicates
            for(int i=0; i<(int) max_peak_locs.size(); ++i)
                heartBeats.push_back(pair<double, int> (max_peak_strengths[i], max_peak_locs[i] + windowStart));
            
            // Calculate the HR for this window
            int windowUpdate = int((1-overlap_ratio)*segment_length+0.5+1e-9);
            if (isFirstSegment) {
                for (int i = 0; i < windowStart; ++i)
                    heartRates.push_back(0);
                isFirstSegment = false;
            }
            
            int count = 0;
            for (int i = 0, sz = (int)segment.size(); i < sz; ++i)
                if (segment[i] > NaN && segment[i] < INFINITY)
                    ++count;
            double rate = (double) max_peak_locs.size() / count * fr;
            
            for(int i = windowStart; i < windowStart+windowUpdate; ++i)
                heartRates.push_back(rate);
            
            windowStart = windowStart + windowUpdate;
            
        }
        
        //Prune the beats counted to include only unique ones
        heartBeats = unique_stable(heartBeats);
        
        //Calc the avg HR for the whole stream
        
        double avg_hr=0;
        if(!heartBeats.empty()){
            //avg_hr = round((double)heartBeats.size() / ((double)heartRates.size() - firstSample) * fr * 60);
            int cnt=0;
            for(int i=firstSample-1; i<(int)temporal_mean.size(); ++i)
                if(temporal_mean[i] != NaN) ++cnt;
            if(cnt==0) avg_hr = 0;
            else
                avg_hr = round((double)heartBeats.size() / cnt * fr * 60);
        }
        
        debug.avg_hr = avg_hr;
        debug.heartBeats = heartBeats;
        debug.heartRates = heartRates;
        
        vector<int> locations;
        for (int i = 0, sz = (int)heartBeats.size(); i < sz; ++i)
            locations.push_back(heartBeats[i].second);
        sort(locations.begin(), locations.end());
        
        return locations;
    }
}